code_generation/code_gen.py

```diff
@@ -117,17 +117,45 @@ def render_dataset_class_maps(self, template_path: Path, data_path: Path, output
 
         self.render_template(template_path=template_path, output_path=output_path, all_map=all_map)
 
+    def render_metadata_types(self, template_path: Path, data_path: Path, output_path: Path):
+        with open(data_path) as data_file:
+            json_data = data_file.read()
+        dataset_meta_data: List[DatasetMapData] = AllDatasetMapData.schema().loads(json_data).all_datasets
+
+        dataset_types = []
+        components = []
+        for dataset in dataset_meta_data:
+            if dataset.is_template:
+                prefixes = ["sym_", "asym_"]
+            else:
+                prefixes = [""]
+            for prefix in prefixes:
+                dataset_types.append(f"{prefix}{dataset.name}")
+
+                if dataset.name == "input":
+                    for component in dataset.components:
+                        components.append(component.names)
+
```

Comment:
It feels like complexity here is more than necessary.

Correct me if im wrong, but wouldnt all components simply be:
`components = [name for components in dataset_meta_data["all_datasets"]["input"]["components"] for name in components["names"]]`

And 
`data_types = [dataset["name"] for dataset in dataset_meta_data["all_datasets"]]`

Reply:
`data_types = [dataset["name"] for dataset in dataset_meta_data["all_datasets"]]` is incorrect as `dataset_meta_data["all_datasets"]]` doesn't include all datatypes. `sym_output` and `asym_output` come attached to `dataset.is_template` (this is how the json config. files are structured).

Making it a comprehension can also be done, but since the first loop is going to be used for components as well and the comprehension would be 'large', I think it is reasonable to keep it this way.

Reply:
see my other component: this may be useful for multiple languages we might ever want to generate an API for

Reply:
> `data_types = [dataset["name"] for dataset in dataset_meta_data["all_datasets"]]` is incorrect as `dataset_meta_data["all_datasets"]]` doesn't include all datatypes. `sym_output` and `asym_output` come attached to `dataset.is_template` (this is how the json config. files are structured).
> 
> Making it a comprehension can also be done, but since the first loop is going to be used for components as well and the comprehension would be 'large', I think it is reasonable to keep it this way.

I meant to point that if the goal is to get all components, only  reading the "input" must be enough if im not mistaken. And not sym_output and asm_output. But its fine either way!